src/sc/sidechain.cpp

```diff
@@ -773,45 +797,119 @@ void CSidechain::UpdateScFees(const CScCertificateView& certView)
                 break;
 
             const auto& entry = scFees.front();
-            LogPrint("sc", "%s():%d - popping f=%d, m=%d from list\n",
+            LogPrint("sc", "%s():%d - popping f=%d, m=%d from list, as scFees maxsize has been reached\n",
                 __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee);
             scFees.pop_front();
-            delta--;
+            --delta;
+        }
+    }
+    // v2 sidechains
+    else {
+        auto scFeeIt = std::find_if(scFees_v2.begin(), scFees_v2.end(), 
+            [&blockHeight](Sidechain::ScFeeData_v2 scFeeElem){return scFeeElem.submissionHeight == blockHeight;});
+
+        // We have not found a scFeeData for the current height, so we add a new one and perform all the checks
+        // on the container size / element age
+        if (scFeeIt == scFees_v2.end()) {
+            scFees_v2.emplace_back(Sidechain::ScFeeData_v2(ftScFee, mbtrScFee, blockHeight));
+
+            // as for v1 sidechains
+            int delta = scFees_v2.size() - maxSizeOfScFeesContainers;
+
+            while (delta > 0)
+            {
+                if (scFees_v2.empty())
+                    break;
+
+                const auto& entry = scFees_v2.front();
+                LogPrint("sc", "%s():%d - popping f=%d, m=%d, h=%d from list, as scFees maxsize has been reached\n",
+                    __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee, entry.submissionHeight);
+                scFees_v2.pop_front();
+                --delta;
+            }
+
+            // Also purge all elements submitted within a block that now is too old to be
+            // considered - this also covers ceasable sidechain, as certificate submission
+            // happens at regular intervals
+            auto& entry = scFees_v2.front();
+            while (entry.submissionHeight <= blockHeight - getMaxSizeOfScFeesContainers())
+            {
+                LogPrint("sc", "%s():%d - popping f=%d, m=%d, h=%d from list, as entry is too old\n",
+                    __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee, entry.submissionHeight);
+                scFees_v2.pop_front();
+                entry = scFees_v2.front();
+            }
+        }
+        // On the other hand, if we found a scFeeData element for the current height, we just update it with lower
+        // ft and/or mbtr fee rates. All the checks on the container size and element age have been performed
+        // at the moment of the element insertion, hence their are not required here
+        else {
+            if (ftScFee < scFeeIt->forwardTxScFee)
+                scFeeIt->forwardTxScFee = ftScFee;
+            if (mbtrScFee < scFeeIt->mbtrTxScFee)
+                scFeeIt->mbtrTxScFee = mbtrScFee;
```

Comment:
We have semantics available which match the description

```suggestion
            scFeeIt->forwardTxScFee = std::min(scFeeIt->forwardTxScFee, ftScFee);
            scFeeIt->mbtrTxScFee = std::min(scFeeIt->mbtrTxScFee, mbtrScFee);
```

Reply:
Agree. Fixed in d39a7f376cd683ff02304198fef02b53898b8375